api/datastore/mysql/mysql.go

```diff
@@ -0,0 +1,615 @@
+		}
+	}
+
+	return datastoreutil.NewValidator(pg), nil
+}
+
+/*
+InsertApp ...
+*/
+func (ds *MySQLDatastore) InsertApp(ctx context.Context, app *models.App) (*models.App, error) {
+	var cbyte []byte
+	var err error
+
+	if app == nil {
+		return nil, models.ErrDatastoreEmptyApp
+	}
+
+	if app.Name == "" {
+		return nil, models.ErrDatastoreEmptyAppName
+	}
+
+	if app.Config != nil {
+		cbyte, err = json.Marshal(app.Config)
+		if err != nil {
+			return nil, err
+		}
+	}
+	stmt, err := ds.db.Prepare("INSERT apps SET name=?,config=?")
+
+	_, err = stmt.Exec(app.Name, string(cbyte))
+
+	if err != nil {
+		return nil, models.ErrAppsAlreadyExists
+	}
+
+	return app, nil
+}
+
+/*
+UpdateApp ...
+*/
+func (ds *MySQLDatastore) UpdateApp(ctx context.Context, newapp *models.App) (*models.App, error) {
+	app := &models.App{Name: newapp.Name}
+	err := ds.Tx(func(tx *sql.Tx) error {
+		row := ds.db.QueryRow(`SELECT config FROM apps WHERE name=?`, app.Name)
+
+		var config string
+		if err := row.Scan(&config); err != nil {
+			if err == sql.ErrNoRows {
+				return models.ErrAppsNotFound
+			}
+			return err
+		}
+
+		if config != "" {
+			err := json.Unmarshal([]byte(config), &app.Config)
+			if err != nil {
+				return err
+			}
+		}
+
+		app.UpdateConfig(newapp.Config)
+
+		cbyte, err := json.Marshal(app.Config)
+		if err != nil {
+			return err
+		}
+
+		stmt, err := ds.db.Prepare(`UPDATE apps SET config=? WHERE name=?`)
+
+		res, err := stmt.Exec(string(cbyte), app.Name)
+
+		if err != nil {
+			return err
+		}
+
+		if n, err := res.RowsAffected(); err != nil {
+			return err
+		} else if n == 0 {
+			return models.ErrAppsNotFound
+		}
+		return nil
+	})
+
+	if err != nil {
+		return nil, err
+	}
+
+	return app, nil
+}
+
+/*
+RemoveApp ...
+*/
+func (ds *MySQLDatastore) RemoveApp(ctx context.Context, appName string) error {
+	_, err := ds.db.Exec(`
+	  DELETE FROM apps
+	  WHERE name = ?
+	`, appName)
+
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+/*
+GetApp ...
+*/
+func (ds *MySQLDatastore) GetApp(ctx context.Context, name string) (*models.App, error) {
+	row := ds.db.QueryRow(`SELECT name, config FROM apps WHERE name=?`, name)
+
+	var resName string
+	var config string
+	err := row.Scan(&resName, &config)
+
+	res := &models.App{
+		Name: resName,
+	}
+
+	json.Unmarshal([]byte(config), &res.Config)
+
+	if err != nil {
+		if err == sql.ErrNoRows {
+			return nil, models.ErrAppsNotFound
+		}
+		return nil, err
+	}
+
+	return res, nil
+}
+
+func scanApp(scanner rowScanner, app *models.App) error {
+	var configStr string
+
+	err := scanner.Scan(
+		&app.Name,
+		&configStr,
+	)
+
+	json.Unmarshal([]byte(configStr), &app.Config)
+
+	return err
+}
+
+/*
+GetApps ...
+*/
+func (ds *MySQLDatastore) GetApps(ctx context.Context, filter *models.AppFilter) ([]*models.App, error) {
+	res := []*models.App{}
+	filterQuery, args := buildFilterAppQuery(filter)
+	rows, err := ds.db.Query(fmt.Sprintf("SELECT DISTINCT * FROM apps %s", filterQuery), args...)
```

Comment:
Oh I just meant to include the literal column names in the query, so that `scanApp` doesn't depend on the set or order of the columns returned by `*`.